lib/activerecord-bitemporal/scope.rb

```diff
@@ -56,11 +56,15 @@ def each_operatable_node(nodes = predicates, &block)
                 case node
                 when Arel::Nodes::LessThan, Arel::Nodes::LessThanOrEqual, Arel::Nodes::GreaterThan, Arel::Nodes::GreaterThanOrEqual
                   y << node if node && node.left.respond_to?(:relation)
-                when Arel::Nodes::Or, Arel::Nodes::And
+                when Arel::Nodes::Or
                   if Gem::Version.new("6.1.0") <= ActiveRecord.version
                     each_operatable_node(node.left) { |node| y << node }
                     each_operatable_node(node.right) { |node| y << node }
                   end
+                when Arel::Nodes::And
+                  if Gem::Version.new("6.1.0") <= ActiveRecord.version
+                    each_operatable_node(node.children) { |node| y << node }
```

Comment:
In `Arel::Nodes::And`, `node.left` `node.right` was not enough, so `node.children` is referenced.
see: https://github.com/rails/rails/blob/v7.0.3.1/activerecord/lib/arel/nodes/and.rb#L13-L19

Reply:
Question: Why do we need to traverse `node.children` only when `6.1.0 < ActiveRecord.version`?

I understand that the `node.left` and `node.right` in `Arel::Nodes::And` can only traverse up to the first two elements of all child nodes, resulting in the `bitemporal_clause` not returning the correct where clause and generating an incorrect query.

However, that should also cause issues with versions prior to 6.0. what is the difference?

Reply:
Also, if that is the case, it should be issues always occur when there are 3 or more conditions in a where clause, but it seems to occur only in the case of JOIN + OR clauses. Why is this?

Reply:
I investigated this.
In Rails 6.0, `Arel::Nodes::Node` `include Enumerable`, but this has been removed since `Rails 6.1` (in Rails 6.0 it was nested `node` iteration).
see: https://github.com/rails/rails/commit/fc38ff6e4417295c870f419f7c164ab5a7dbc4a5#diff-6c6d1e7194d82128058d81887cc94c39dc67b74ca57a95c58a272771f7b1ec3cL9

This changes the behavior of [`Array(nodes)`](https://github.com/kufu/activerecord-bitemporal/blob/cd0cc56e9ae742d29a46c15fb08c282a9cf159ad/lib/activerecord-bitemporal/scope.rb#L55).

```ruby
users = Arel::Table.new(:users)
node = users[:name].eq("homu").or(users[:name].eq("mami"))

# In Rails 6.0 `node.to_a` is called, in Rails 6.1 it is just `[node]`.
pp Array(node).count
# 6.0 => 14
# 6.1 => 1
```

This was broken in `Rails 6.1` because of this.

Reply:
use superclass to support other nodes as much as possible.
https://github.com/kufu/activerecord-bitemporal/pull/99/commits/c667e843f58342d92c8ec93e6bcb48055b8b696c

Reply:
Makes sense. We need a stable conditions visitor...

I also found [other nodes](https://github.com/rails/rails/blob/v6.1.7/activerecord/lib/arel/nodes/update_statement.rb) which can contain where clauses, but since these are not created by `bitemporal_default_scope`, I understood that we doesn't need to traverse these nodes (is this right?)

Reply:
> Makes sense. We need a stable conditions visitor...

yep, I understand...

> I also found [other nodes](https://github.com/rails/rails/blob/v6.1.7/activerecord/lib/arel/nodes/update_statement.rb) which can contain where clauses, but since these are not created by bitemporal_default_scope, I understood that we doesn't need to traverse these nodes (is this right?)

Yes, `UpdateStatement` is used for `ActiveRecord::Relation#update_all` and so on.
https://github.com/rails/rails/blob/v6.1.6.1/activerecord/lib/arel/update_manager.rb#L9-#L9
https://github.com/rails/rails/blob/v6.1.6.1/activerecord/lib/active_record/relation.rb#L446-#L446